client/TracyProfiler.cpp

```diff
@@ -2907,8 +2915,195 @@ void Profiler::HandleSymbolQuery( uint64_t symbol )
+    while( it != mappings.end() && it->start_address < end_address ) {
+        if (it->start_address > remainder_start_address) {
+            // The current mapping starts after the remainder of the address
+            // range that we needs to cover, so some of that address range
+            // isn't covered by any mapping.
+            return false;
+        }
+        remainder_start_address = it->end_address;
+        ++it;
+    }
+    // Some of the address range isn't covered by any mapping.
+    if (remainder_start_address < end_address) return false;
+    *end_iterator = it;
+    if (*end_iterator < *start_iterator) {
+        abort();
+    }
+    if (*end_iterator > (*start_iterator) + 1) {
+        abort();
+    }
+    return true;
+}
+                      
+
+// Attempts to make the specified `mapping` readable if it isn't already.
+// Returns true if and only if the mapping is readable.
+bool EnsureReadable( MappingInfo* mapping )
+{
+    if( mapping->perm_r )
+    {
+        // The mapping is already readable.
+        return true;
+    }
+    if( mapping->mprotect_fails )
+    {
+        // We have already see mprotect failing on this mapping.
+        return false;
+    }
+    int prot = PROT_READ;
+    if( mapping->perm_w ) prot |= PROT_WRITE;
+    if( mapping->perm_x ) prot |= PROT_EXEC;
+    if( mprotect( reinterpret_cast<void*>( mapping->start_address ),
+                 mapping->end_address - mapping->start_address, prot ) == -1 )
+    {
+        // Failed to make the mapping readable. Update `mapping`
+        // so the next call will be fast.
+        mapping->mprotect_fails = true;
+        return false;
+    }
+    // The mapping is now readable. Update `mapping` so the next call will be fast.
+    mapping->perm_r = true;
+    return true;
+}
+
+// Attempts to make the specified address range readable.
+bool EnsureReadable( std::uintptr_t start_address, std::uintptr_t end_address )
+{
+    // Lock ensuring reentrancy of this function.
+    static std::mutex s_mutex;
+    const std::lock_guard<std::mutex> lock( s_mutex );
+    // Static state managed by this function. Not constant, we mutate that state as
+    // we turn some mappings readable. Initially parsed once here, updated as needed below.
+    static MappingInfoVector mappings = ParseMappings();
+    MappingInfoIterator start_iter, end_iter;
+    if (!GetMappingsRange(mappings, start_address, end_address, &start_iter, &end_iter))
+    {
+        // Some of this address range isn't in any known mapping. Try parsing agains, maybe
+        // mappings changed.
+        mappings = ParseMappings();
+        if (!GetMappingsRange(mappings, start_address, end_address, &start_iter, &end_iter))
+        {
+            // Still not mapped? This shouldn't happen...
+            return false;
+        }
+    }
+    if (end_iter < start_iter) {
+        abort();
+    }
+    int k = 0;
+    for (auto it = start_iter; it != end_iter; ++it)
```

Comment:
I don't think it would be possible to have one symbol span multiple memory mappings?

Reply:
ah, i was wondering about that. thinking more about it, i agree, and it does result in a great simplification.